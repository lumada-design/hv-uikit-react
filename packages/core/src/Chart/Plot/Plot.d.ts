import * as React from "react";
import { StandardProps } from "@material-ui/core";
// eslint-disable-next-line import/no-unresolved
import { Config, Layout, PlotData as PlotlyPlotData } from "plotly.js";

export type PlotData = PlotlyPlotData;
export type PlotLayout = Layout;
export type PlotConfig = Config;

export type HvPlotClassKey = "";

export interface HvPlotProps
  extends StandardProps<React.HTMLAttributes<HTMLDivElement>, HvPlotClassKey> {
  /**
   * Plotly data object.
   */
  data: PlotData[];
  /**
   * Plotly layout object.
   */
  layout?: PlotLayout;
  /**
   * Plotly config object.
   */
  config?: PlotConfig;
  /**
   * OnHover function.
   */
  onHover?: (data: PlotData) => void;
  /**
   * OnUnHover function.
   */
  onUnHover?: (data: PlotData) => void;
  /**
   * Function to be call after plot render.
   */
  afterPlot?: (data: PlotData, ref: any) => void;
  /**
   * Plot revision
   */
  revision?: number;
}

export default function HvPlot(props: HvPlotProps): JSX.Element | null;
