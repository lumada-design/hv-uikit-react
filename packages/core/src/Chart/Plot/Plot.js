/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { memo, useLayoutEffect, useRef, useState } from "react";
import PropTypes from "prop-types";
import Plotly from "plotly.js-basic-dist";
import createPlotlyComponent from "react-plotly.js/factory";

/**
 * Plotly plot.
 *
 * @param classes
 * @param data
 * @param layoutclasses
 * @param onHover
 * @param onUnHover
 * @returns {*}
 * @constructor
 */
const Plot = ({ data, layout, config, onHover, onUnHover, afterPlot }) => {
  const [revision, setRevision] = useState(0);
  const [newData, setNewData] = useState(data);

  const PlotGraph = createPlotlyComponent(Plotly);

  const ref = useRef(null);

  const afterPlotInternal = () => {
    const afterData = afterPlot(newData, ref);
    if (afterData) {
      setNewData(afterData);
      setRevision(revision + 1);
    }
  };

  /**
   * Call in the first render.
   */
  if (afterPlot) {
    useLayoutEffect(afterPlotInternal, []);
  }

  return (
    <div ref={ref}>
      <PlotGraph
        data={newData}
        layout={layout}
        config={config}
        revision={revision}
        onHover={eventData => onHover(eventData)}
        onUnhover={eventData => onUnHover(eventData)}
        onAfterPlot={afterPlot && afterPlotInternal}
        style={{ position: "relative" }}
      />
    </div>
  );
};

Plot.propTypes = {
  /**
   * Plotly data object.
   */
  data: PropTypes.arrayOf(PropTypes.instanceOf(Object)).isRequired,
  /**
   * Plotly layout object.
   */
  layout: PropTypes.instanceOf(Object).isRequired,
  /**
   * Plotly config object.
   */
  config: PropTypes.instanceOf(Object).isRequired,
  /**
   * OnHover function.
   */
  onHover: PropTypes.func,
  /**
   * OnUnHover function.
   */
  onUnHover: PropTypes.func,
  /**
   * Function to be call after plot render.
   */
  afterPlot: PropTypes.func
};

Plot.defaultProps = {
  onHover: null,
  onUnHover: null,
  afterPlot: undefined
};

const arePropsEqual = (prevProps, nextProps) =>
  prevProps.data === nextProps.data && prevProps.layout === nextProps.layout;

export default memo(Plot, arePropsEqual);
