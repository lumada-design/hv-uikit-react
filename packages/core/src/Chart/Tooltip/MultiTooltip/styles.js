const styles = (theme) => ({
  root: {
    backgroundColor: theme.hv.palette.atmosphere.atmo1,
    width: "fit-content",
    boxShadow: theme.hv.shadows[1],
    zIndex: 100,
  },
  title: {
    padding: `15px ${theme.hv.spacing.sm}px`,
    borderBottom: `3px solid ${theme.hv.palette.atmosphere.atmo2}`,
  },
  valuesContainer: {
    padding: `${theme.hv.spacing.sm}px`,
  },
  values: {
    display: "flex",
    alignItems: "center",
    paddingBottom: `${theme.hv.spacing.sm}px`,
    "&:last-child": {
      paddingBottom: 0,
    },
  },
  color: {
    width: `${theme.hv.spacing.xs}px`,
    height: `${theme.hv.spacing.xs}px`,
  },
  separator: {
    width: `${theme.hv.spacing.xs}px`,
  },
  separatorColor: {
    width: "5px",
  },
});

export default styles;
