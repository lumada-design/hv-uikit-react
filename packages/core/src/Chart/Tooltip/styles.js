const styles = {
  root: {
    position: "fixed",
    zIndex: 100,
    pointerEvents: "none",
  },
};

export default styles;
