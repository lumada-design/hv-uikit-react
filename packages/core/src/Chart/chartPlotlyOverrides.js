/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Set the property if it is null.
 * @param object
 * @param property
 * @param value
 */
const setterIfNil = (object, property, value) => {
  // eslint-disable-next-line no-param-reassign
  object[property] = object[property] || value;
};

/**
 * Auxiliary functions to set the layout in order to follow the DS guideline.
 *
 */

export const setYaxis = (inputLayout, styles, isHorizontal, yAxisTitle) => {
  const layout = inputLayout;

  setterIfNil(layout, "yaxis", {});

  setterIfNil(layout.yaxis, "automargin", true);

  setterIfNil(layout.yaxis, "title", {});

  setterIfNil(layout.yaxis.title, "font", {
    family: styles.vizText.fontFamily,
    size: styles.vizText.fontSize,
    color: styles.vizText.color
  });

  setterIfNil(layout.yaxis, "fixedrange", true);

  setterIfNil(layout.yaxis, "tickcolor", styles.lineColor);
  setterIfNil(layout.yaxis, "ticklen", 2);
  setterIfNil(layout.yaxis, "tickfont", {
    family: styles.vizText.fontFamily,
    size: styles.vizText.fontSize,
    color: styles.vizText.color
  });

  setterIfNil(layout.yaxis, "gridcolor", styles.gridColor);
  setterIfNil(layout.yaxis, "showgrid", !isHorizontal);

  setterIfNil(layout.yaxis, "linecolor", styles.lineColor);
  setterIfNil(layout.yaxis, "linewidth", 1);
  setterIfNil(layout.yaxis, "zeroline", false);

  if (yAxisTitle) {
    setterIfNil(layout, "yaxis", {});
    setterIfNil(layout.yaxis, "title", {});
    setterIfNil(layout.yaxis.title, "text", yAxisTitle);
  }

  return layout;
};

export const setXaxis = (inputLayout, styles, xAxisTitle, isHorizontal) => {
  const layout = inputLayout;
  setterIfNil(layout, "xaxis", {});

  setterIfNil(layout.xaxis, "automargin", true);

  setterIfNil(layout.xaxis, "title", {});

  setterIfNil(layout.xaxis.title, "font", {
    family: styles.vizText.fontFamily,
    size: styles.vizText.fontSize,
    color: styles.vizText.color
  });

  setterIfNil(layout.xaxis, "fixedrange", true);

  setterIfNil(layout.xaxis, "tickcolor", styles.lineColor);
  setterIfNil(layout.xaxis, "ticklen", 2);
  setterIfNil(layout.xaxis, "tickfont", {
    family: styles.vizText.fontFamily,
    size: styles.vizText.fontSize,
    color: styles.vizText.color
  });

  setterIfNil(layout.xaxis, "gridcolor", styles.gridColor);
  setterIfNil(layout.xaxis, "showgrid", isHorizontal);

  setterIfNil(layout.xaxis, "linewidth", 1);
  setterIfNil(layout.xaxis, "linecolor", styles.lineColor);
  setterIfNil(layout.xaxis, "zeroline", false);

  if (xAxisTitle) {
    setterIfNil(layout, "xaxis", {});
    setterIfNil(layout.xaxis, "title", {});
    setterIfNil(layout.xaxis.title, "text", xAxisTitle);
  }
  return layout;
};

export const setLegend = (inputLayout, styles) => {
  const layout = inputLayout;
  setterIfNil(layout, "legend", {
    x: 1,
    y: 1.1,
    xanchor: "right",
    yanchor: "bottom",
    orientation: "h"
  });

  setterIfNil(layout.legend, "font", {
    family: styles.vizText.fontFamily,
    size: styles.vizText.fontSize,
    color: styles.vizText.color
  });
  return layout;
};

export const setLayout = (inputLayout, styles) => {
  const layout = inputLayout;

  setterIfNil(layout, "colorway", Object.values(styles.defaultColors));

  setterIfNil(layout, "autosize", true);

  setterIfNil(layout, "margin", {
    l: 50,
    b: 50,
    t: 10,
    pad: 0
  });

  setterIfNil(layout, "plot_bgcolor", styles.plotColor);

  setterIfNil(layout, "paper_bgcolor", styles.plotColor);
};
