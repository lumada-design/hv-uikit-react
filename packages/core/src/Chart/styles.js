/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const styles = theme => ({
  root: {
    padding: "40px",
    backgroundColor: theme.hv.palette.atmosphere.atmo1
  },
  titleContainer: {
    backgroundColor: theme.hv.palette.atmosphere.atmo1
  },
  subtitle: {
    marginTop: `${theme.hv.spacing.xs}px`
  },
  defaultColors: {
    ...theme.hv.viz.palette.categorical
  },
  vizText: {
    ...theme.hv.typography.vizText,
    fontFamily: theme.hv.typography.fontFamily,
    fontSize: 10
  },
  lineColor: theme.hv.palette.atmosphere.atmo6,
  gridColor: theme.hv.palette.atmosphere.atmo5,
  paddingTop: {
    paddingTop: "20px"
  },
  plotColor: theme.hv.palette.atmosphere.atmo1
});

export default styles;
