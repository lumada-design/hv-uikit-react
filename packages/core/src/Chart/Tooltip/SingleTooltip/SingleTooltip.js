import React, { memo } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core";
import Typography from "../../../Typography";
import styles from "./styles";

const SingleTooltip = ({ classes, title, value }) => (
  <div className={classes.root}>
    <div>
      <Typography variant="highlightText">{title}</Typography>
    </div>
    <div className={classes.separator} />
    <div>
      <Typography>{value}</Typography>
    </div>
  </div>
);

SingleTooltip.propTypes = {
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root class.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the separator between title and value.
     */
    separator: PropTypes.string,
  }).isRequired,
  /**
   * Title of the chart.
   */
  title: PropTypes.string.isRequired,
  /**
   * Value of the chart.
   */
  value: PropTypes.number.isRequired,
};

const arePropsEqual = (prevProps, nextProps) =>
  prevProps.title === nextProps.title && prevProps.value === nextProps.value;

export default withStyles(styles, { name: "HvChartSingleTooltip" })(
  memo(SingleTooltip, arePropsEqual)
);
