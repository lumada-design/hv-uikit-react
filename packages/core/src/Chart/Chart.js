/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";
import isNil from "lodash/isNil";
import classNames from "classnames";
import Tooltip from "./Tooltip";
import {
  setLayout,
  setLegend,
  setXaxis,
  setYaxis
} from "./chartPlotlyOverrides";
import styleCreator from "./styles";
import Typography from "../Typography";
import Plot from "./Plot";

/**
 * Setter of default layout properties.
 *
 * @param inputLayout
 * @param theme
 * @returns {*}
 */
const propsLayoutSetter = (
  inputLayout,
  theme,
  isHorizontal,
  xAxisTitle,
  yAxisTitle
) => {
  const styles = styleCreator(theme);
  const layout = inputLayout === undefined ? {} : inputLayout;

  // Layout
  setLayout(layout, styles);

  // Legend
  setLegend(layout, styles);

  // Xaxis
  setXaxis(layout, styles, xAxisTitle, isHorizontal);

  // Yaxis
  setYaxis(layout, styles, yAxisTitle, isHorizontal);

  return layout;
};

/**
 * Component responsible for the presentation of the barchart component.
 *
 * @param title
 * @param subtitle
 * @param data
 * @param layout
 * @param config
 * @param tooltipType
 * @returns {*}
 * @constructor
 */
const Chart = ({
  classes,
  theme,
  title,
  subtitle,
  data,
  layout,
  config,
  tooltipType,
  afterPlot,
  xAxisTitle,
  yAxisTitle
}) => {
  const [isHover, setIsHover] = useState(false);
  const [coordinates, setCoordinates] = useState({ x: 0, y: 0 });
  const [dataTooltip, setDataTooltip] = useState();

  // Check if the barchart is horizontal or vertical.
  const isHorizontal = !isNil(data[0].orientation)
    ? data[0].orientation.toUpperCase() === "H"
    : false;

  const newLayout = propsLayoutSetter(
    layout,
    theme,
    isHorizontal,
    xAxisTitle,
    yAxisTitle
  );

  /**
   * Extract data from the plotly onHover event to be used to create the tooltip.
   *
   * @param eventData
   */
  const onHover = eventData => {
    const dataFromPoints = {
      title: isHorizontal ? eventData.points[0].y : eventData.points[0].x,
      elements: []
    };
    eventData.points.forEach(p => {
      dataFromPoints.elements.push({
        color: p.fullData.marker
          ? p.fullData.marker.color
          : p.fullData.line.color,
        name: p.fullData.name,
        value: isHorizontal ? p.x : p.y
      });
    });

    setDataTooltip(dataFromPoints);

    if (!isHover) setIsHover(true);
  };

  const onUnHover = () => setIsHover(false);

  /**
   * Obtains the cursor coordinates to send to the tooltip.
   * @param eventData
   */
  const onMouseMove = eventData => {
    setCoordinates({
      x: eventData.pageX,
      y: eventData.pageY
    });
  };

  const useSingle = tooltipType === "single";

  return (
    <div classes={classes}>
      {isHover && (
        <Tooltip
          coordinates={coordinates}
          data={dataTooltip}
          useSingle={useSingle}
        />
      )}
      <div className={classes.root}>
        <div className={classes.titleContainer}>
          {title && <Typography variant="mTitle">{title}</Typography>}
          <div className={classes.subtitle}>
            {subtitle && <Typography variant="infoText">{subtitle}</Typography>}
          </div>
        </div>
        <div
          className={classNames({ [classes.paddingTop]: title })}
          onMouseMove={e => onMouseMove(e)}
        >
          <Plot
            title={title}
            data={data}
            layout={newLayout}
            config={{ responsive: true, displayModeBar: false, ...config }}
            onHover={onHover}
            onUnHover={onUnHover}
            afterPlot={afterPlot}
          />
        </div>
      </div>
    </div>
  );
};

Chart.propTypes = {
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.instanceOf(Object),
  /**
   * Theme.
   */
  theme: PropTypes.instanceOf(Object).isRequired,
  /**
   * Title of the chart.
   */
  title: PropTypes.string,
  /**
   * Subtitle of the chart.
   */
  subtitle: PropTypes.string,
  /**
   * Plotly data object (see https://plot.ly/javascript/reference/).
   */
  data: PropTypes.arrayOf(PropTypes.instanceOf(Object)).isRequired,
  /**
   * Plotly layout object (see https://plot.ly/javascript/reference/#layout).
   */
  layout: PropTypes.instanceOf(Object).isRequired,
  /**
   * Plotly config object (see https://plot.ly/javascript/configuration-options/).
   */
  config: PropTypes.instanceOf(Object),
  /**
   * Defines if should use a single or multiline tooltip.
   */
  tooltipType: PropTypes.oneOf(["single", "multiple"]),
  /**
   * Function to be called after plot render.
   */
  afterPlot: PropTypes.func,
  /**
   * Defines the X axis title.
   */
  xAxisTitle: PropTypes.string,
  /**
   * Defines the Y axis title.
   */
  yAxisTitle: PropTypes.string
};

Chart.defaultProps = {
  classes: null,
  title: "",
  subtitle: "",
  tooltipType: "multiple",
  config: null,
  afterPlot: undefined,
  xAxisTitle: null,
  yAxisTitle: null
};

export default Chart;
