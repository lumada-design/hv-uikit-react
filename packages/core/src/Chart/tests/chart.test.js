/* eslint-env jest */

import React from "react";
import { mount } from "enzyme";

import HvProvider from "../../Provider";
import Chart from "..";
import Plot from "../Plot";
import Tooltip from "../Tooltip";
import SingleTooltip from "../Tooltip/SingleTooltip";
import MultiTooltip from "../Tooltip/MultiTooltip";

describe("Chart withStyles", () => {
  let wrapper;

  const trace1 = {
    x: [2300, 1000, 8500],
    y: ["January", "February", "March"],
    name: "Sales Target",
    type: "bar",
    orientation: "h",
    hoverinfo: "none",
  };

  const data = [trace1];
  const layout = {};

  beforeEach(() => {
    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Chart data={data} layout={layout} />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(Chart)).toMatchSnapshot();
  });

  it("should render the Chart", () => {
    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Chart data={data} layout={layout} />
      </HvProvider>
    );
    const chart = wrapper.find(Chart);
    expect(chart.length).toBe(1);
  });

  it("should render the Plot", () => {
    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Chart data={data} layout={layout} />
      </HvProvider>
    );
    const plot = wrapper.find(Plot);
    expect(plot.length).toBe(1);
  });
});

describe("Tooltip withStyles", () => {
  let wrapper;
  const data = { title: "" };
  const coordinates = { x: 0, y: 0 };

  beforeEach(() => {
    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Tooltip data={data} coordinates={coordinates} />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(Tooltip)).toMatchSnapshot();
  });

  it("should render a singleTooltip", () => {
    const coord = { x: 1, y: 2 };
    const singleData = { title: "title", elements: [{ value: 22 }] };

    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Tooltip data={singleData} coordinates={coord} useSingle />
      </HvProvider>
    );

    const singleTooltip = wrapper.find(SingleTooltip);
    expect(singleTooltip.length).toBe(1);
  });

  it("should render a multiTooltip", () => {
    const coord = { x: 1, y: 2 };
    const singleData = { title: "title", elements: [{ name: "name", value: 22 }] };

    wrapper = mount(
      <HvProvider disableCssBaseline>
        <Tooltip data={singleData} coordinates={coord} />
      </HvProvider>
    );

    const multiTooltip = wrapper.find(MultiTooltip);
    expect(multiTooltip.length).toBe(1);
  });
});
