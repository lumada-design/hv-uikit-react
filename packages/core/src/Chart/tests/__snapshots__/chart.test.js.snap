// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart withStyles should render correctly 1`] = `
<WithStyles(Chart)
  data={
    Array [
      Object {
        "hoverinfo": "none",
        "name": "Sales Target",
        "orientation": "h",
        "type": "bar",
        "x": Array [
          2300,
          1000,
          8500,
        ],
        "y": Array [
          "January",
          "February",
          "March",
        ],
      },
    ]
  }
  layout={Object {}}
/>
`;

exports[`Tooltip withStyles should render correctly 1`] = `
<WithStyles(MainToolTip)
  coordinates={Object {}}
  data={Object {}}
/>
`;
